src/server/editorServices.ts

```diff
@@ -1492,47 +1718,111 @@ namespace ts.server {
+                ([path, isRootOfInferredProject]) => {
+                    if (!ignoreIfNotRootOfInferredProject || isRootOfInferredProject) {
+                        const info = this.getScriptInfoForPath(path as Path);
+                        Debug.assert(!!info);
+                        return info;
+                    }
+                }
+            );
+            this.reloadConfiguredsProjectForFiles(openFiles, /*delayReload*/ true);
+            this.delayInferredProjectsRefresh();
+        }
+
+        /**
+         * This function goes through all the openFiles and tries to file the config file for them.
+         * If the config file is found and it refers to existing project, it reloads it either immediately
+         * or schedules it for reload depending on delayReload option
+         * If the there is no existing project it just opens the configured project for the config file
+         */
+        private reloadConfiguredsProjectForFiles(openFiles: ScriptInfo[], delayReload: boolean) {
+            const updatedProjects = createMap<true>();
             // try to reload config file for all open files
-            for (const info of this.openFiles) {
-                this.openOrUpdateConfiguredProjectForFile(info.fileName);
+            for (const info of openFiles) {
+                // This tries to search for a tsconfig.json for the given file. If we found it,
+                // we first detect if there is already a configured project created for it: if so,
+                // we re- read the tsconfig file content and update the project only if we havent already done so
+                // otherwise we create a new one.
+                const configFileName = this.getConfigFileNameForFile(info);
+                if (configFileName) {
+                    const project = this.findConfiguredProjectByProjectName(configFileName);
+                    if (!project) {
+                        this.createConfiguredProject(configFileName, info.fileName);
+                        updatedProjects.set(configFileName, true);
+                    }
+                    else if (!updatedProjects.has(configFileName)) {
+                        if (delayReload) {
+                            project.pendingReload = true;
+                            this.delayUpdateProjectGraph(project);
+                        }
+                        else {
+                            this.reloadConfiguredProject(project);
+                        }
+                        updatedProjects.set(configFileName, true);
+                    }
+                }
+            }
+        }
+
+        /**
+         * Remove the root of inferred project if script info is part of another project
+         */
+        private removeRootOfInferredProjectIfNowPartOfOtherProject(info: ScriptInfo) {
+            // If the script info is root of inferred project, it could only be first containing project

```

Comment:
Hmm check if this helps. I can explain it further if this is not still clear
```TypeScript
            // If the script info is root of inferred project, it could only be first containing project
            // since info is added as root to the inferred project only when there are no other projects containing it
            // So when it is root of the inferred project and after project structure updates its now part
            // of multiple project it needs to be removed from that inferred project because:
            // - references in inferred project supercede the root part
            // - root / reference in non - inferred project beats root in inferred project

            // eg. say this is structure /a/b/a.ts /a/b/c.ts where c.ts references a.ts
            // When a.ts is opened, since there is no configured project/external project a.ts can be part of
            // a.ts is added as root to inferred project.
            // Now at time of opening c.ts, c.ts is also not aprt of any existing project,
            // so it will be added to inferred project as a root. (for sake of this example assume single inferred project is false)
            // So at this poing a.ts is part of first inferred project and second inferred project (of which c.ts is root)
            // And hence it needs to be removed from the first inferred project.

```